Implement desktop notifications.

YuriSolovyov opened the pull request:
This is another attempt to add desktop notifications based on work of @inikulin in #25.
The main feedback in original PR was about notification content and action we perform on notification click. 

With minimal changes, I was able to make it look like so:
![image](https://cloud.githubusercontent.com/assets/2168518/13728530/0e041fb2-e92d-11e5-974e-c7f0b58319fb.png)

Though I'd like to purpose a bit different style:
![image](https://cloud.githubusercontent.com/assets/2168518/13729060/8a6cecfa-e93c-11e5-84ca-e1ff402ac0c6.png)

With all that there are still some questions left:
- what do we show if there are more than 1 notification?
  - list notification? 

  ![image](https://developer.chrome.com/static/images/list_notification.png)
 - just say "there are N notifications for you" ? or "there is `foo` and N-1 other notifications for you"?

- we need to somehow track last notification timestamp, so we don't show same thing several times.
- on > 1 notification, where should click bring you?

Fixes #12 
Fixes #70 

/cc @sindresorhus

YuriSolovyov commented:
Idea: maybe if there > 1 notifications, we can add button like `Next` so user can cycle through them and `Mark all as read` to mark all at once?

sindresorhus commented:
> Though I'd like to purpose a bit different style:

Yeah. I like that style better.

> what do we show if there are more than 1 notification?

Shouldn't there be a Chrome notification for each GitHub notification?

YuriSolovyov commented:
> Shouldn't there be a Chrome notification for each GitHub notification?

I thought that could be quite annoying for active github users.

YuriSolovyov commented:
@sindresorhus ATM the only problem is to convert [GitHub API](https://developer.github.com/v3/activity/notifications/) url to the actual issue url.

YuriSolovyov commented:
For now, I found 2 ways of getting issue url of notification:

1 - We can get if from notification info, but it is hackish and not future proof:
```js
[apiRes.repository.html_url].concat(apiRes.subject.url.split('/').slice(-2)).join('/')
// -> "https://github.com/octocat/Hello-World/issues/123"
```

2 - We can make a request to GitHub API to get this info:
```
curl 'https://api.github.com/repos/octokit/octokit.rb/issues/123'
//     ^- this is same as apiRes.subject.url
```
gives:
```js
{
  "url": "https://api.github.com/repos/octokit/octokit.rb/issues/123",
  "repository_url": "https://api.github.com/repos/octokit/octokit.rb",
  "labels_url": "https://api.github.com/repos/octokit/octokit.rb/issues/123/labels{/name}",
  "comments_url": "https://api.github.com/repos/octokit/octokit.rb/issues/123/comments",
  "events_url": "https://api.github.com/repos/octokit/octokit.rb/issues/123/events",
  "html_url": "https://github.com/octokit/octokit.rb/pull/123",
  "id": 6130680,
  "number": 123,
  "title": "Implement repo assignees",
  // ...
}
```
and we need `html_url` from here.
I think it is ok to make this request when user clicks notification, so we send request lazily when needed.

YuriSolovyov commented:
@sindresorhus this is ready for review.

Narno commented:
Hello @YuriSolovyov, I'm testing your fork: feedback next week! :-)

YuriSolovyov commented:
Cool!
/cc @fstoerkle @gcochard

sindresorhus commented on extension/main.js:
```diff
@@ -1,6 +1,68 @@
+	function handleNotificationClicked(notificationId) {
+		const url = window.GitHubNotify.settings.get(notificationId);
+		if (url) {
+			fetch(url).then(res => {
```
```js
fetch(url).then(res => res.json())
```

sindresorhus commented on extension/manifest.json:
```diff
@@ -9,7 +9,8 @@
 	"permissions": [
-		"alarms"
+		"alarms",
+		"notifications"
```
Shouldn't this be an optional permission?

YuriSolovyov commented:
@sindresorhus addressed feedback.
I noticed that even though [we are setting `optional_permissions`](https://github.com/sindresorhus/notifier-for-github-chrome/blob/master/extension/main.js#L120), this flag is never checked, should I remove it or make it work somehow?

Narno commented:
@YuriSolovyov it's a trick to avoid popin each time user refuse the new permission (true and false are !== undefined).

YuriSolovyov commented:
@Narno I guess I should add same thing for notifications?

Narno commented:
Yes, but if you find a class solution, that would be cool! ;-)

YuriSolovyov commented:
@Narno I don't quite get how/if user can grant this permission later on in this case...

Narno commented:
@YuriSolovyov yep, the only way to do that for the moment is to re-install the extension. I don't know if it's a real problem or not :-)

YuriSolovyov commented:
@Narno 
For now I just factored out this to Promise-based API.
We might want to add a setting like `"Allow querying my tabs"` or something

YuriSolovyov commented:
I'm also thinking about making use of [`reason` field](https://developer.github.com/v3/activity/notifications/#notification-reasons) - we can have a map of messages for every type of `reason` field, so we can add this info to notification like 
```
<message>
<repo>
New comment | You were mentioned | You were assigned | You are subsribed
```
... and so on
cc @Narno

YuriSolovyov commented:
Now I'm also fixing #70 here

YuriSolovyov commented:
Can anyone try 
```bash
curl -v -H "Authorization: token <token>" "https://api.github.com/notifications?per_page=1"
```
and tell me if it returns non-empty ETag header?

Matsv commented:
@YuriSolovyov curl output: http://pastebin.com/8a3iU4g2

YuriSolovyov commented:
@Matsv please replace `<token>` with real token, you can generate one in settings, make sure you added `notifications` permission

Matsv commented:
@YuriSolovyov Ok, one sec

Matsv commented:
@YuriSolovyov http://pastebin.com/dWJYnMAM with notications from the public-apis repo

YuriSolovyov commented:
Thanks, @Matsv, seems like I'm not the only one.
Ok, I sent a letter to Github, maybe I missed something.

YuriSolovyov commented:
linter is complaining about `snake_case`, but since it comes from rails app, it is kind of normal and we can't change it, so we have (at least) 2 options:
1. Somehow disable the rule atomically, just for 1 function
2. Replace obj with `Map`:
```js
// object:
const reasons = {
  subscribed:   'You are watching the repository',
  manual:       'You are subscribed to this thread',
  author:       'You created this thread',
  comment:      'New comment',
  mention:      'You were mentioned',
  team_mention: 'Your team was mentioned',
  state_change: 'Thread status changed',
  assign:       'You were assigned to the issue',
  default:      ''
};

// Map:
const reasons = new Map([
  ['subscribed',   'You are watching the repository'  ]
  ['manual',       'You are subscribed to this thread']
  ['author',       'You created this thread',         ]
  ['comment',      'New comment'                      ]
  ['mention',      'You were mentioned'               ]
  ['team_mention', 'Your team was mentioned'          ]
  ['state_change', 'Thread status changed'            ]
  ['assign',       'You were assigned to the issue'   ]
  ['default':      ''                                 ]
]);
```

Any ideas?

Also, I'd like to get some feedback about wording of these messages, maybe someone has better ideas, or it is fine for now?

gcochard commented:
You can always disable it with /* eslint-disable snake_case */ before and then /* eslint-enable snake_case */ at the end of your handling code.

YuriSolovyov commented:
This is ready for another round of review.

gcochard commented on package.json:
```diff
@@ -3,7 +3,7 @@
   },
   "devDependencies": {
-    "chrome-stub": "^1.2.0",
+    "sinon-chrome": "^1.1.2",
```
Just curious, why this change? It appears the usage is exactly the same in the tests. Are you planning on adding more to the tests that chrome-stub doesn't support?

gcochard commented on extension/api.js:
```diff
@@ -64,67 +35,105 @@
-		if (!token) {
-			cb(new Error('missing token'), null, window.GitHubNotify.defaults.interval);
-			return;
+		if (/api.github.com\/$/.test(rootUrl)) {
```
Should continue to use `(api\.)?github\.com` in the regex to support users who simply have `github.com` as their root URL.

gcochard replied:
Nevermind. Scrolled down and see that it was moved from below. This should work as-is.

gcochard commented on extension/api.js:
```diff
@@ -64,67 +35,105 @@
-			url += 'api/v3/notifications';
+		const tabUrl = `${rootUrl}notifications`;
+		if (window.GitHubNotify.settings.get('useParticipatingCount')) {
+			return `${tabUrl}/participating`;
```
Drop the slash here.

YuriSolovyov commented:
@gcochard you are right about stub changes, works ok with `chrome-stub`, I just remember there were some problems with it (can't remember what), anyway, I switched back to `chrome-stub`.

YuriSolovyov commented:
Btw, I don't think we need to stub XMLHttpRequest anymore...

YuriSolovyov commented:
Btw, there is a bit weird implementation detail about reason field in API response that can make things a bit confusing:
> Note that the reason is modified on a per-thread basis, and can change, if the reason on a later notification is different.

>For example, if you are the author of an issue, subsequent notifications on that issue will have a reason of author. If you're then mentioned on the same issue, the notifications you fetch thereafter will have a reason of mention. The reason remains as mention, regardless of whether you're ever mentioned again.

gcochard commented:
I think that also applies to the emails github sends out and users should expect that behavior. Once github fixes it, we get the fix for free.

If you do encounter any other issues with `chrome-stub`, let me know by filling an issue on it. I'm the author and I wrote it primarily for testing this extension so I have an interest in making it work for all of our use cases.

gcochard commented:
LGTM

YuriSolovyov commented:
I'd like to use it for a couple more days just to be sure

Narno commented:
Hello @YuriSolovyov, I'm (re)testing your branch since yesterday: notifications never appears, may be because the extension don't asks me for notifications permissions.

YuriSolovyov commented:
@Narno anything interesting in console?

YuriSolovyov commented:
@Narno can you check following things:
1. does API response contain `Last-Modified` header? if not there will be `null` in `lastModifed` key in local storage (something is very weird here: I had this header on work, but now it is absent on home PC)
2. notifications might not some just after upgrade, but at least on 2nd check they should come.
3. I'm not sure, but it seems like on linux notifications only appear when browser is focused, might be distro-dependent.

YuriSolovyov commented:
3. nah, appear just fine even without focus

sindresorhus commented:
ping @Narno :)

YuriSolovyov commented:
Status update: I was pretty busy over past couple weeks, but I'm going to back to it this weekend, there is indeed a bug here and I'm going to fix it.

YuriSolovyov commented:
@Narno I updated PR, should be better now, I'll use for couple more days and post final go/non-go later.

Narno commented:
Nice, I will try the new version this night!

Narno commented:
OK, now Chrome asks me the right thing when I checks the "Show desktop notifications" option. Good new!

Narno commented:
OK, after several tests, in different contexts, desktop notifications works really fine! \o/

YuriSolovyov commented:
@Narno can you recall if transition from version without notifications to this PR was ok?
Anything you did by hand that should've happen automatically?
Or switched just fine?

Narno commented:
Hello @YuriSolovyov, what I have tested:
* installing the PR extension in // of the current extension (with the same token): OK
* uninstalling the current extension then installing the PR extension (with the same token) : OK
* installing the PR extension on a fresh Chrome installation (with a new token) : OK

Done on my laptop computer (Mac OS X Yosemite) and on my desktop computer (Windows Seven).

YuriSolovyov commented:
@sindresorhus I think this is ready.

YuriSolovyov commented:
@sindresorhus now it's your turn :smile:

Narno commented:
:+1:

sindresorhus commented:
Awesome! Superb work @YuriSolovyov :)

Could you add something to the readme about the desktop notification feature?

I've published a new version. Should be live within an hour.

YuriSolovyov commented:
@sindresorhus sure, I'll add a couple words.
